AppKit/CPTextView/CPTypesetter.j

```diff
@@ -0,0 +1,467 @@
+
+/*
+ *  CPTypesetter.j
+ *  AppKit
+ *
+ *  Created by Daniel Boehringer on 27/12/2013.
+ *  All modifications copyright Daniel Boehringer 2013.
+ *  Based on original work by
+ *  Emmanuel Maillard on 27/02/2010.
+ *  Copyright Emmanuel Maillard 2010.
+ *
+ *  FIXME: paragraphStyle indent information is currently not properly respected
+ *         collect all run heights per line for proper baseline alignment
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+@import <Foundation/CPObject.j>
+@import "CPParagraphStyle.j"
+@import "CPTextStorage.j"
+
+@global _isNewlineCharacter
+
+// forward declare these classes for type matching
+@class CPLayoutManager
+@class CPTextContainer
+@class CPTextView
+
+/*
+    CPTypesetterControlCharacterAction
+*/
+CPTypesetterZeroAdvancementAction = 1 << 0;
+CPTypesetterWhitespaceAction      = 1 << 1;
+CPSTypesetterHorizontalTabAction  = 1 << 2;
+CPTypesetterLineBreakAction       = 1 << 3;
+CPTypesetterParagraphBreakAction  = 1 << 4;
+CPTypesetterContainerBreakAction  = 1 << 5;
+
+var _measuringContext,
+    _measuringContextFont,
+    _isCanvasSizingInvalid,
+    _didTestCanvasSizingValid,
+    _sharedSimpleTypesetter,
+    _sizingCache;
+
+function _widthOfStringForFont(aString, aFont)
+{
```

Comment:
@Dogild sizeWithFont: turned out to slow for CPTextView. The function _widthOfStringForFont is is ~10x faster on Chrome and FF (where it can be canvas-based). In Safari, it falls back to sizeWithFont: since canvas text sizing is broken there. Please also see my comment in issue #1400